src/tests/GotHttpAdapter.test.ts

```diff
@@ -0,0 +1,362 @@
+        contentType,
+      });
+
+      expect(mockGot.post).toHaveBeenCalledWith(expect.anything(), expect.objectContaining({
+        [expectedKey]: body,
+      }));
+    });
+
+    it('should return the body from the response', async () => {
+      const response = {
+        body: {},
+        headers: {},
+      };
+      mockGot.post.mockResolvedValue(response);
+
+      const result = await httpAdapter.post('http://example.com');
+      expect(result).toEqual(response.body);
+    });
+
+    it('should return the full response when this is requested', async () => {
+      const response = {
+        body: {},
+        headers: {},
+      };
+      mockGot.post.mockResolvedValue(response);
+
+      const result = await httpAdapter.post('http://example.com', {}, {}, { resolveFullResponse: true });
+      expect(result).toEqual(response);
+    });
+
+    it.each([
+      429,
+      401,
+      404,
+      500,
+    ])('should throw HttpStatusCodeError[%s] when got throws HTTPError', async (statusCode: number) => {
+      const httpError = produceFoolInstance(HTTPError);
+      httpError.message = 'Http Error';
+      httpError.response = {
+        statusCode,
+      } as HTTPError['response'];
+      mockGot.post.mockRejectedValueOnce(httpError);
+
+      let caughtErr;
+      try {
+        await httpAdapter.post('http://example.com');
+      } catch (err) {
+        caughtErr = err;
+      }
+
+      expect(caughtErr).toBeInstanceOf(HttpStatusCodeError);
+      expect(caughtErr.getStatusCode()).toEqual(statusCode);
+    });
+
+    it('should throw HttpRequestError when got throws RequestError', async () => {
+      const requestError = produceFoolInstance(RequestError);
+      requestError.message = 'Request Error';
+      requestError.name = 'RequestError';
+      mockGot.post.mockRejectedValueOnce(requestError);
+
+      let caughtErr;
+      const url = 'http://example.com';
+      try {
+        await httpAdapter.post(url);
+      } catch (err) {
+        caughtErr = err;
+      }
+
+      expect(caughtErr).toBeInstanceOf(HttpRequestError);
+      expect(caughtErr.request).toEqual({
+        url,
+        method: 'POST',
+        headers: {},
+        body: {},
+      });
+    });
+
+    it('should throw HttpGenericError when got throws subclass of RequestError', async () => {
+      const parseError = produceFoolInstance(ParseError);
+      parseError.message = 'Unexpected token < at position...';
+      parseError.name = 'ParseError';
+      mockGot.post.mockRejectedValueOnce(parseError);
+
+      let caughtErr;
+      const url = 'http://example.com';
+      try {
+        await httpAdapter.post(url);
+      } catch (err) {
+        caughtErr = err;
+      }
+
+      expect(caughtErr).toBeInstanceOf(HttpGenericError);
+      expect(caughtErr.originalError).toEqual(parseError);
+    });
+  });
+});
+
+interface Constructable {
+  new(...args: any[]): any
+}
+
+type Mutable<T> = { -readonly [P in keyof T ]: T[P] };
```

Comment:
```suggestion
type OptionalProperties<T extends Constructable> = Partial<{ [P in keyof InstanceType<T>]: InstanceType<T>[P] }>;

function produceFoolInstance<T extends Constructable>(Class: T, properties?: OptionalProperties<T>): InstanceType<T> {
  const object = Object.setPrototypeOf({}, Class.prototype);
  return Object.assign(object, properties);
}
```

Example usage:

```typescript
      const httpError = produceFoolInstance(HTTPError, {
        message: 'Http Error',
        response: {
          statusCode,
        } as HTTPError['response'],
      });
```

Reply:
I like the approach, replaced Partial with DeepPartial in order to remove the `as Whatever` casting for objs.